domain/network/state/subnet.go

```diff
@@ -0,0 +1,426 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/network"
+)
+
+const (
+	subnetTypeBase              = 0
+	subnetTypeFanOverlaySegment = 1
+)
+
+// AddSubnet creates a subnet.
+func (st *State) AddSubnet(
+	ctx context.Context,
+	uuid utils.UUID,
+	cidr string,
+	providerID network.Id,
+	providerNetworkID network.Id,
+	VLANTag int,
+	availabilityZones []string,
+	spaceUUID string,
+	fanInfo *network.FanCIDRs,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var subnetType int
+	if fanInfo != nil {
+		subnetType = subnetTypeFanOverlaySegment
+	}
+	var spaceUUIDValue any
+	spaceUUIDValue = spaceUUID
+	if spaceUUID == "" {
+		spaceUUIDValue = nil
+	}
+	pnUUID, err := utils.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertSubnetStmt := `
+INSERT INTO subnet (uuid, cidr, vlan_tag, space_uuid, subnet_type_id)
+VALUES (?, ?, ?, ?, ?)`
+	insertSubnetAssociationStmt := `
+INSERT INTO subnet_association (subject_subnet_uuid, associated_subnet_uuid, association_type_id)
+VALUES (?, ?, 0)` // For the moment the only allowed association is 'overlay_of' and therefore its ID is hard-coded here.
+	retrieveUnderlaySubnetUUIDStmt := `
+SELECT uuid
+FROM   subnet
+WHERE  cidr = ?`
+	insertSubnetProviderIDStmt := `
+INSERT INTO provider_subnet (provider_id, subnet_uuid)
+VALUES (?, ?)`
+	insertSubnetProviderNetworkIDStmt := `
+INSERT INTO provider_network (uuid, provider_network_id)
+VALUES (?, ?)`
+	insertSubnetProviderNetworkSubnetStmt := `
+INSERT INTO provider_network_subnet (provider_network_uuid, subnet_uuid)
+VALUES (?, ?)`
+	insertAvailabilityZoneStmt := `
```

Comment:
What happens if two subnets are in the same AZ? 

You'll get a key violation when attempting to add the AZ a second time.

Reply:
fixed in de1aeee9c36f7d4a5160b36298e93861650bedaa